### PR TITLE
[CAD-336] Use of the switchboard after shutdown.

### PR DESCRIPTION
**Pull request by ksaric**
https://jira.iohk.io/browse/CAD-336

### CODE REVIEW

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs [hunk 1]
```diff
@@ -63,20 +63,31 @@ We are using an |MVar| because we spawn a set of backends that may try to send m
...
     , sbLogBuffer :: !(Cardano.BM.Backend.LogBuffer.LogBuffer a)
     , sbLogBE     :: !(Cardano.BM.Backend.Log.Log a)
     , sbBackends  :: NamedBackends a
+    , sbRunning   :: !SwitchboardStatus
     }
 
 type NamedBackends a = [(BackendKind, Backend a)]
 
+-- | A bool isomorphic status for the switchboard
+-- which we can use.
```
Comment (CodiePP) [inline iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: what does "which we can use" mean? the comment is not needed IMO.

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs [hunk 2]
```diff
@@ -124,8 +135,14 @@ instance IsEffectuator Switchboard a where
                     else atomically $ TBQ.writeTBQueue q i
 
         sb <- readMVar (getSB switchboard)
-        writequeue (sbQueue sb) item
 
+        -- First we check to see if we have the switchboard running.
+        -- If the switchboard is running, we write to the queue,
+        -- otherwise we report the error on the stderr.
```
Comment (CodiePP) [inline iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: the code is self-explanatory

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs [hunk 3]
```diff
@@ -226,29 +244,55 @@ instance (FromJSON a, ToJSON a) => IsBackend Switchboard a where
...
+            , sbLogBE = katipBE
+            , sbBackends = bs
+            , sbRunning = SwitchboardRunning
+            }
 
         return sb
 
     unrealize switchboard = do
-        let clearMVar :: MVar some -> IO ()
+        -- How is this "clearing" anything?
```
Comment (CodiePP) [inline iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: would you want to rename it?

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs [hunk 4]
```diff
@@ -226,29 +244,55 @@ instance (FromJSON a, ToJSON a) => IsBackend Switchboard a where
...
+        let clearMVar :: MVar (SwitchboardInternal a) -> IO ()
             clearMVar = void . tryTakeMVar
 
+        -- This is kind of weird. This function is generally used for
+        -- the modification of the internal state, not breaking the encapsulation.
         (dispatcher, queue) <- withMVar (getSB switchboard) (\sb -> return (sbDispatch sb, sbQueue sb))
```
Comment (CodiePP) [inline iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: right. this should better get a lock on the MVar, send the kill pill and then clean up. or, was the idea to give all the backends a last chance to use the switchboard?

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs [hunk 5]
```diff
@@ -226,29 +244,55 @@ instance (FromJSON a, ToJSON a) => IsBackend Switchboard a where
...
+            flagSwitchboardStopped sb = 
+                SwitchboardInternal 
+                    { sbQueue = sbQueue sb
+                    , sbDispatch = sbDispatch sb
+                    , sbLogBuffer = sbLogBuffer sb
+                    , sbLogBE = sbLogBE sb
+                    , sbBackends = sbBackends sb
+                    , sbRunning = SwitchboardNotRunning
+                    }
+
```
Comment (CodiePP) [inline iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: why not: `flagSwitchboardStopped sb = sb { sbRunning = SwitchboardNotRunning }`

Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: really clearing/releasing the components?

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs (hunk 3 above)
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: I would remove it. I don't see what this does.

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs (hunk 2 above)
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: A lot of things are self-explanatory until you need to figure out what they do. A couple of lines of comments go a long way.

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs (hunk 1 above)
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: Meaning we can use it as a bool isomorphic status.

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs (hunk 4 above)
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: Exactly, this should be done _inside_ the `withMVar`.
I don't see how this gives all the backends any chance in either case.

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs (hunk 5 above)
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: That is the same, but less explicit.

File: iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs (hunk 1 above)
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs]: This _can_ be removed if you think it's pointless.

**Comment by ksaric** [comment]
@CodiePP Tested on the node (`mainnet`) and through GHCI, works correctly.
An issue was raised in the #haskell-best-practices Slack channel for the confirmation of the issue I run into.

**Pull request merged**